core/src/apps/cardano/sign_tx.py

```diff
@@ -67,522 +84,355 @@
     from apps.common.paths import PathSchema
 
-    CborizedTokenBundle = dict[bytes, dict[bytes, int]]
-    CborizedTxOutput = tuple[bytes, Union[int, tuple[int, CborizedTokenBundle]]]
-    CborizedSignedTx = tuple[dict, dict, Optional[cbor.Raw]]
-    TxHash = bytes
 
 MINTING_POLICY_ID_LENGTH = 28
 MAX_ASSET_NAME_LENGTH = 32
-MAX_TX_CHUNK_SIZE = 256
 
 
 @seed.with_keychain
 async def sign_tx(
     ctx: wire.Context, msg: CardanoSignTx, keychain: seed.Keychain
-) -> CardanoSignedTx:
-    if msg.fee > LOVELACE_MAX_SUPPLY:
-        raise wire.ProcessError("Fee is out of range!")
+) -> CardanoSignTxFinished:
+    is_network_id_verifiable = await _validate_tx_signing_request(ctx, msg)
 
-    validate_network_info(msg.network_id, msg.protocol_magic)
+    builder = TxBuilder(keychain, msg)
+
+    common_input_account = await _process_inputs(ctx, msg, keychain, builder)
+    total_amount_to_show = await _process_outputs(
+        ctx, msg, keychain, builder, common_input_account
+    )
+    builder.add_fee()
+    builder.add_ttl()
+    await _process_certificates(ctx, msg, keychain, builder)
+    await _process_withdrawals(ctx, msg, builder)
+    catalyst_signature = await _process_auxiliary_data(ctx, msg, keychain, builder)
+    builder.add_validity_interval_start()
+
+    if msg.signing_mode == CardanoTxSigningMode.ORDINARY_TRANSACTION:
+        await confirm_transaction(
+            ctx,
+            total_amount_to_show,
+            msg.fee,
+            msg.protocol_magic,
+            msg.ttl,
+            msg.validity_interval_start,
+            is_network_id_verifiable,
+        )
+    elif msg.signing_mode == CardanoTxSigningMode.POOL_REGISTRATION:
+        await confirm_stake_pool_registration_final(ctx)
 
     try:
-        if _has_stake_pool_registration(msg):
-            cborized_tx, tx_hash = await _sign_stake_pool_registration_tx(
-                ctx, msg, keychain
+        tx_hash = builder.get_tx_hash()
+
+        response: CardanoTxItemAck | CardanoTxAuxiliaryDataSupplement | CardanoTxWitnessResponse = (
+            CardanoTxItemAck()
+            if catalyst_signature is None
+            else CardanoTxAuxiliaryDataSupplement(
+                type=CardanoTxAuxiliaryDataSupplementType.CATALYST_REGISTRATION_SIGNATURE,
+                catalyst_signature=catalyst_signature,
             )
-        else:
-            cborized_tx, tx_hash = await _sign_ordinary_tx(ctx, msg, keychain)
-
-        signed_tx_chunks = cbor.encode_chunked(cborized_tx, MAX_TX_CHUNK_SIZE)
-
-        for signed_tx_chunk in signed_tx_chunks:
-            response = CardanoSignedTxChunk(signed_tx_chunk=signed_tx_chunk)
-            await ctx.call(response, CardanoSignedTxChunkAck)
+        )
 
-        return CardanoSignedTx(tx_hash=tx_hash, serialized_tx=None)
+        for _ in range(msg.witnesses_count):
+            witness_request = await ctx.call(response, CardanoTxWitnessRequest)
+            response = CardanoTxWitnessResponse(
```

Comment:
This is the policy Ledger has: https://github.com/vacuumlabs/ledger-app-cardano-shelley/blob/25357a944067d3fd234931e190bc706423906676/src/securityPolicy.c#L395 so I'd make sure to replicate it here as well

It warns for non-standard (i.e. "reasonable") account number and address index over a certain threshold value:

https://github.com/vacuumlabs/ledger-app-cardano-shelley/blob/965d37dca65c6c8f3b2ee6471eb9ef9c4a2f6933/src/bip44.c#L9

More importantly, it rejects other than staking key witnesses for the case when registering a stake pool as an owner (otherwise a user may be tricked by the operator to unknowingly confirm spending their own funds)

Reply:
The "reasonable" part can be achieved by using `SCHEMA_ADDRESS` and `SCHEMA_STAKING` schemas. The stake pool owner parts needs to be covered separately.

Reply:
Discussed offline - will be added later on.